codegen/src/main/scala/bsp/codegen/MarkdownRenderer.scala

```diff
@@ -0,0 +1,116 @@
+package bsp.codegen
+
+import software.amazon.smithy.model.shapes.ShapeId
+import bsp.codegen.JsonRPCMethodType.Notification
+import bsp.codegen.JsonRPCMethodType.Request
+import bsp.codegen.Hint.Documentation
+import cats.syntax.all._
+import scala.collection.mutable.{Set => MSet}
+
+object MarkdownRenderer {
+  def render(tree: DocTree): String = {
+    val state = MSet.empty[ShapeId]
+    val renderer = new MarkdownRenderer(state)
+    renderer.render(tree).get.mkString(System.lineSeparator())
+  }
+}
+
+class MarkdownRenderer private (state: MSet[ShapeId]) {
+
+  import dsl._
+
+  def render(tree: DocTree): Lines = {
```

Comment:
Just an idea, it might be nice to make a `Renderer` trait with a single abstract `render` if that's the only thing that's really exposed. For no other reason than to cement the pattern of having a renderer that exposes a render just to make it clear in the future when we want to create another one, we just create a new renderer and implement `render`.